notebooks/tutorials/cloud-ml-engine/Model training and prediction with Cloud Machine Learning Engine.ipynb

```diff
@@ -0,0 +1,571 @@
+    "+ Exports the model with the [Python `pickle` library](https://docs.python.org/3/library/pickle.html)\n",
+    "\n",
+    "The following model training code is not executed within this notebook. Instead, it is saved to a Python file and packaged as a Python module that runs on Cloud ML Engine after you submit the training job."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%writefile ./census_training/train.py\n",
+    "import argparse\n",
+    "import pickle\n",
+    "import pandas as pd\n",
+    "\n",
+    "from google.cloud import storage\n",
+    "\n",
+    "from sklearn.ensemble import RandomForestClassifier\n",
+    "from sklearn.feature_selection import SelectKBest\n",
+    "from sklearn.pipeline import FeatureUnion\n",
+    "from sklearn.pipeline import Pipeline\n",
+    "from sklearn.preprocessing import LabelBinarizer\n",
+    "\n",
+    "parser = argparse.ArgumentParser()\n",
+    "parser.add_argument(\"--bucket-name\", help=\"The bucket name\", required=True)\n",
+    "\n",
+    "arguments, unknown = parser.parse_known_args()\n",
+    "bucket_name = arguments.bucket_name\n",
+    "\n",
+    "# Define the format of your input data, including unused columns.\n",
+    "# These are the columns from the census data files.\n",
+    "COLUMNS = (\n",
+    "    'age',\n",
+    "    'workclass',\n",
+    "    'fnlwgt',\n",
+    "    'education',\n",
+    "    'education-num',\n",
+    "    'marital-status',\n",
+    "    'occupation',\n",
+    "    'relationship',\n",
+    "    'race',\n",
+    "    'sex',\n",
+    "    'capital-gain',\n",
+    "    'capital-loss',\n",
+    "    'hours-per-week',\n",
+    "    'native-country',\n",
+    "    'income-level'\n",
+    ")\n",
+    "\n",
+    "# Categorical columns are columns that need to be turned into a numerical value\n",
+    "# to be used by scikit-learn\n",
+    "CATEGORICAL_COLUMNS = (\n",
+    "    'workclass',\n",
+    "    'education',\n",
+    "    'marital-status',\n",
+    "    'occupation',\n",
+    "    'relationship',\n",
+    "    'race',\n",
+    "    'sex',\n",
+    "    'native-country'\n",
+    ")\n",
+    "\n",
+    "# Create a Cloud Storage client to download the census data\n",
+    "storage_client = storage.Client()\n",
+    "\n",
+    "# Download the data\n",
+    "public_bucket = storage_client.bucket('cloud-samples-data')\n",
+    "blob = public_bucket.blob('ml-engine/sklearn/census_data/adult.data')\n",
+    "blob.download_to_filename('adult.data')\n",
+    "\n",
+    "# Load the training census dataset\n",
+    "with open(\"./adult.data\", \"r\") as train_data:\n",
+    "    raw_training_data = pd.read_csv(train_data, header=None, names=COLUMNS)\n",
+    "    # Remove the whitespaces in categorical features\n",
+    "    for col in CATEGORICAL_COLUMNS:\n",
+    "        raw_training_data[col] = raw_training_data[col].apply(\n",
+    "            lambda x: str(x).strip())\n",
```

Comment:
Had a conversation with @happyhuman about this. Essentially:

This line that strips whitespace becomes part of the scikit-learn pipeline that performs the same preprocessing on the raw input data. Therefore there's no need to do further preprocessing on the input (test) data when submitting online predictions. However, if you do further preprocessing to the test data before sending the prediction request, it may throw off the predictions.

So (further down in this file) we'll need to use the data directly from our Cloud Storage bucket instead of the ten example instances from the docs. (It's an exact copy of the source data from UCI Machine Learning Repository, without any additional preprocessing.)